### PR TITLE
7.x XML forms tests

### PR DESCRIPTION
**Pull request by qadan**
This includes a suite of tests for XML Forms, including basic form management, association functionality, and some tests integrated with the Test Solution Pack.

One thing to note: the third commit down there slightly changes the way the Import Form form validates imported XML. This change was the best option to get the Import Form tests working, as SimpleTest's internal browser was uploading XML with an application/octet-stream mimetype.

This depends on the pull request at https://github.com/Islandora/islandora/pull/454

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -12,16 +12,25 @@ env:
 before_install:
   - cd $HOME
   - git clone -b 7.x git://github.com/Islandora/islandora.git
+  - git clone -b 7.x git://github.com/Islandora/php_lib.git
+  - git clone -b 7.x git://github.com/Islandora/objective_forms.git
+  - git clone -b 7.x git://github.com/discoverygarden/islandora_solution_pack_test.git
```
Comment (willtp87) [inline .travis.yml]: An Islandora repo shouldn't depend on a community repo.

File: .travis.yml [hunk 2]
```diff
@@ -12,16 +12,25 @@ env:
...
 script:
-  #commented out jenkins because there is no build.xml file.  TODO. create build.xml
-  #- ant -buildfile sites/all/modules/islandora_xml_forms/build.xml lint
+  # TODO: fix this so it ignores appropriate files.
+  # - ant -buildfile sites/all/modules/islandora_xml_forms/build.xml lint
   - $ISLANDORA_DIR/tests/scripts/line_endings.sh sites/all/modules/islandora_xml_forms
-  - drush coder-review --reviews=production,security,style,i18n,potx,sniffer islandora_xml_forms
-  #The copy, paste detector is commented out because it was reconizing whole classes as duplicate lines. TODO fix this
-  #- phpcpd --names *.module,*.inc,*.test sites/all/modules/islandora_xml_forms
+  # TODO: oh my goodness this is a mess. someday, someone's going to have to clean this out.
```
Comment (willtp87) [inline .travis.yml]: This comment is less than precise.

File: .travis.yml [hunk 3]
```diff
@@ -12,16 +12,25 @@ env:
...
+  # - drush coder-review --reviews=production,security,style,i18n,potx,sniffer xml_forms
```
Comment (willtp87) [inline .travis.yml:32]: Why are we disabling coder?

File: .travis.yml [hunk 4]
```diff
@@ -12,16 +12,25 @@ env:
...
   - cd $HOME/drupal-*
   - ln -s $TRAVIS_BUILD_DIR sites/all/modules/islandora_xml_forms
-  - drush en --yes --user=1 islandora_xml_forms
+  - ln -s $HOME/php_lib sites/all/modules/php_lib
+  - ln -s $HOME/objective_forms sites/all/modules/objective_forms
+  - ln -s $HOME/islandora_solution_pack_test sites/all/modules/islandora_solution_pack_test
+  - ln -s $HOME/islandora_solution_pack_collection sites/all/modules/islandora_solution_pack_collection
+  - drush en --yes --user=1 xml_form_builder xml_form_elements islandora_basic_collection islandora_test_cm
 script:
-  #commented out jenkins because there is no build.xml file.  TODO. create build.xml
```
Comment (willtp87) [inline .travis.yml:21]: Is this todo no longer valid?

File: builder/Create.inc [hunk 5]
```diff
@@ -86,12 +86,10 @@ function xml_form_builder_create_validate_name(array $element, array &$form_stat
...
-    $file_is_xml = $_FILES['files']['type']['file'] == 'text/xml';
-    if ($file_is_xml) {
-      module_load_include('inc', 'xml_form_api', 'XMLFormDefinition');
-      $filename = $_FILES['files']['tmp_name']['file'];
-      $definition = new DOMDocument();
-      $definition->load($filename);
+    module_load_include('inc', 'xml_form_api', 'XMLFormDefinition');
+    $filename = $_FILES['files']['tmp_name']['file'];
+    $definition = new DOMDocument();
+    if (@$definition->load($filename)) {
```
Comment (willtp87) [inline builder/Create.inc]: @ is almost always a BAD idea you may squash things unintentionally. I would advise handling errors instead, was something especially broken about how this was working before?

File: tests/islandora_xml_forms_associate.test [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+    // Disable an association! The only association, actually.
+    $this->drupalGet('admin/islandora/xmlform/Test Solution Pack MODS Form/associations');
+    // Before we get going, let's make sure we won't encounter a million errors.
+    if (!$this->assertLink('Disable')) {
+      return;
+    }
+    $disable_links = $this->xpath('//a[normalize-space(text())=Disable]');
+    if (isset($disable_links[1])) {
+      // http://www.youtube.com/watch?v=sqcLjcSloXs
```
Comment (willtp87) [inline tests/islandora_xml_forms_associate.test]: no

File: tests/islandora_xml_forms_associate.test [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+    $this->clickLink("Disable");
+    $this->drupalGet("islandora/object/$collection_pid/manage/overview/ingest");
+    $this->assertText("Select file to upload", "Collection with no form associations skips directly to file upload step.", 'Islandora');
+
+    // Create a new form association with several customizations.
+    $dsid = $this->randomName();
+    $custom_transform_filename = $this->randomName() . ".xsl";
+    $self_transform_filename = $this->randomName() . ".xsl";
+    file_put_contents(DRUPAL_ROOT . "/sites/all/modules/islandora_xml_forms/builder/transforms/" . $custom_transform_filename, file_get_contents(drupal_get_path('module', 'islandora_test_cm') . "/xsl/mods_to_dc_custom.xsl"));
+    file_put_contents(DRUPAL_ROOT . "/sites/all/modules/islandora_xml_forms/builder/self_transforms/" . $self_transform_filename, file_get_contents(drupal_get_path('module', 'islandora_test_cm') . "/xsl/self_transform.xsl"));
```
Comment (willtp87) [inline tests/islandora_xml_forms_associate.test]: "/sites/all/modules/islandora_xml_forms/builder/self_transforms/$self_transform_filename" --some in other places

File: tests/islandora_xml_forms_associate.test [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+    // Let's check out the metadata for this new object!
+    if ($this->assertDatastreams($object_pid, array($dsid, 'DC'))) {
+
+      // MODS metadata.
+      $object_xml = $this->curlExec(array(
+        CURLOPT_URL => url("islandora/object/$object_pid/datastream/$dsid/download", array('absolute' => TRUE)),
+        CURLOPT_NOBODY => FALSE,
+      ));
+      file_put_contents('/tmp/modsthingy.xml', $object_xml);
```
Comment (willtp87) [inline tests/islandora_xml_forms_associate.test]: Debug :(  --some in other places

File: tests/islandora_xml_forms_associate.test [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+    }
+    $this->clickLink("Disable");
+    $this->drupalGet("islandora/object/$collection_pid/manage/overview/ingest");
+    $this->assertText("Select file to upload", "Collection with no form associations skips directly to file upload step.", 'Islandora');
+
+    // Create a new form association with several customizations.
+    $dsid = $this->randomName();
+    $custom_transform_filename = $this->randomName() . ".xsl";
+    $self_transform_filename = $this->randomName() . ".xsl";
+    file_put_contents(DRUPAL_ROOT . "/sites/all/modules/islandora_xml_forms/builder/transforms/" . $custom_transform_filename, file_get_contents(drupal_get_path('module', 'islandora_test_cm') . "/xsl/mods_to_dc_custom.xsl"));
```
Comment (willtp87) [inline tests/islandora_xml_forms_associate.test]: Should probably use Drupal's file api.  https://api.drupal.org/api/drupal/includes!file.inc/group/file/7  --some in other places

File: tests/islandora_xml_forms_associate.test [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+      'islandora_basic_collection',
+      'xml_form_builder',
+      'islandora_test_cm',
+    ));
+  }
+
+  /**
+   * Test form association functions.
+   */
+  public function testFormAssociationFunctions() {
```
Comment (willtp87) [inline tests/islandora_xml_forms_associate.test]: Test of each function should be its own test.  --some in other places

File: tests/islandora_xml_forms_test_cm_tests.test [hunk 11]
```diff
@@ -0,0 +1,99 @@
+<?php
+
+/**
+ * @file
+ * http://i.imgur.com/xjFp0MR.jpg
```
Comment (willtp87) [inline tests/islandora_xml_forms_test_cm_tests.test]: I enjoy the humor, but we do not control imgur.  This needs to be replaced with something more reliable.

File: builder/Create.inc (hunk 5 above)
Comment (qadan) [inline builder/Create.inc]: The previous iteration was not accepting files that were valid XML - which perhaps even had a valid XML mimetype - but were not necessarily uploaded as text/xml. This was breaking in browsers that uploaded as application/xml, as well as in the SimpleTest internal browser, which wasn't uploading using a valid XML mimetype at all.

As for the @ ... it seemed redundant to have Drupal pass through two messages - one from the failure of DOMDocument::load() and one from form_set_error(). I could try catching the exception instead?

File: tests/islandora_xml_forms_associate.test (hunk 6 above)
Comment (qadan) [inline tests/islandora_xml_forms_associate.test]: ;)

... I'll remove it.

File: .travis.yml (hunk 4 above)
Comment (qadan) [inline .travis.yml:21]: It's partially no longer valid. We're still not doing this line, but the todo should be reworded.

File: .travis.yml (hunk 2 above)
Comment (qadan) [inline .travis.yml]: haha can-do ... I forgot to better-ify this before I pushed.

File: builder/Create.inc (hunk 5 above)
Comment (willtp87) [inline builder/Create.inc]: I would definitely be more comfortable with catching the most specific exception.

File: .travis.yml (hunk 3 above)
Comment (qadan) [inline .travis.yml:32]: Coder wasn't actually running properly up until now; the changes above to before_install fix that, but they also mean that coder actually runs, where it then comes back with a million billion errors. Someone's going to have to either go through and fix all of the errors - which are many - or go through and implement a bunch of ignore comments.

File: .travis.yml (hunk 1 above)
Comment (qadan) [inline .travis.yml]: Should we fork islandora_solution_pack_test to Islandora's repo, then? I'm not quite sure what the process for that is.

Comment (willtp87) [inline .travis.yml]: A review of that module and a transfer of ownership need to take place.

Comment (willtp87) [inline .travis.yml]: Had a talk with @qadan about this, going in a different direction.  It's not right to cause a new dependency for testing.

**Comment by qadan** [comment]
that is a crapload of commits whoopsadaisy

**Pull request merged**